content/blog/2023/2024-01-27-deserialize-with-jsoncreator.md

```diff
@@ -0,0 +1,465 @@
+When we run the test above, we see this error:
+````text
+com.fasterxml.jackson.databind.exc.InvalidDefinitionException: 
+Cannot construct instance of `com.reflectoring.userdetails.model.UserData` 
+(no Creators, like default constructor, exist): 
+cannot deserialize from Object value (no delegate- or property-based Creator)
+ at [Source: (String)"{"id":100,"firstName":"Ranjani","lastName":"Harish",
+ "city":"Sydney","createdDate":"2024-01-16"}"; line: 1, column: 2]
+````
+This is because the `ObjectMapper` class by default looks for a no-arg constructor to set the values. **Since our Java class is immutable, it has neither a no-arg constructor
+nor setter methods to set values**. Also, we see that the `UserData` class has multiple constructors. Therefore, we would need a way to 
+instruct the `ObjectMapper` class to use the right constructor for deserialization.
+Let's modify our code to use the `@JsonCreator` annotation as :
+````java
+@JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public UserData(@JsonProperty("id") long id, 
+                    @JsonProperty("firstName") String firstName,
+                    @JsonProperty("lastName") String lastName, 
+                    @JsonProperty("city") String city) {
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.city = city;
+        this.createdDate = LocalDate.now();
+    }
+````
+Now, when we run our test again, we can see that the deserialization is successful.
+Let's look at the additional annotation we've added to get this working.
+- By applying the `@JsonCreator` annotation to the constructor, the Jackson deserializer knows which constructor needs to be used.
+- `JsonCreator.Mode.PROPERTIES` indicates that the creator should match the incoming object with the constructor arguments. This is the most commonly used JsonCreator Mode.
+- We annotate all the constructor arguments with `@JsonProperty` for the creator to map the arguments.
+
+### Understanding all the available JsonCreator modes:
+- **JsonCreator.Mode.PROPERTIES** : This is the most commonly used mode where every constructor/factory argument is annotated with either `@JsonProperty` to indicate the name of the property to bind to.
+- **JsonCreator.Mode.DELEGATING** : Single-argument constructor/factory method without JsonProperty annotation for the argument. Here, Jackson first binds JSON into type of the argument, and then calls creator. This is often used in conjunction with JsonValue (used for serialization).
+- **JsonCreator.Mode.DEFAULT** : If no mode or DEFAULT mode is chosen, Jackson internally decides which of PROPERTIES / DELEGATING mode to choose from.
+- **JsonCreator.Mode.DISABLED** : This mode indicates the creator method is not to be used.
+
+
+In the further sections, we will take a look at examples and how to use them effectively.
+
+### Additional usecases:
```

Comment:
Let's make this a main section (## Additional usecases) without the colon at the end. Then we do not need the enumeration, but can use titles. So
`**1. Only a single @JsonCreator annotation with the same mode can be applied in the same class**`
will become:
`### Only a single @JsonCreator annotation with the same mode can be applied in the same class`